examples/platforms/samr21/radio.c

```diff
@@ -0,0 +1,611 @@
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the OpenThread platform abstraction for radio communication.
+ *
+ */
+
+#include "asf.h"
+
+#include <openthread/config.h>
+#include <openthread/platform/alarm-milli.h>
+#include <openthread/platform/radio.h>
+
+#include "platform-samr21.h"
+#include "common/logging.hpp"
+
+enum
+{
+    IEEE802154_ACK_LENGTH      = 5,
+    IEEE802154_FCS_SIZE        = 2
+};
+
+static otRadioFrame sTransmitFrame;
+static uint8_t      sTransmitPsdu[OT_RADIO_FRAME_MAX_SIZE];
+
+static otRadioFrame sReceiveFrame;
+
+static bool         sSleep          = false;
+static bool         sRxEnable       = false;
+static bool         sTxDone         = false;
+static bool         sRxDone         = false;
+static otError      sTxStatus       = OT_ERROR_NONE;
+static int8_t       sPower          = OPENTHREAD_CONFIG_DEFAULT_MAX_TRANSMIT_POWER;
+static otRadioState sState          = OT_RADIO_STATE_DISABLED;
+static bool         sPromiscuous    = false;
+static uint8_t      sChannel        = 0xFF;
+
+static int8_t       sMaxRssi;
+static uint32_t     sScanStartTime;
+static uint16_t     sScanDuration;
+static bool         sStartScan      = false;
+
+static int8_t         sTxPowerTable[] = { 4, 4, 3, 3, 3, 2, 1, 0,
+                                          -1, -2, -3, -4, -6, -8, -12, -17
+                                        };
+
+/*******************************************************************************
+ * Static
+ ******************************************************************************/
+
+static void radioSleep()
+{
+    if (!sSleep)
+    {
+        PHY_SetRxState(false);
+        PHY_Sleep();
+
+        sSleep = true;
+        sRxEnable = false;
+    }
+}
+
+static void radioWakeup()
+{
+    if (sSleep)
+    {
+        PHY_Wakeup();
+    }
+}
+
+static void radioRxEnable()
+{
+    if (sSleep)
+    {
+        PHY_Wakeup();
+
+        sSleep = false;
+    }
+
+    if (!sRxEnable)
+    {
+        PHY_SetRxState(true);
+
+        sRxEnable = true;
+    }
+}
+
+static void radioTrxOff()
+{
+    if (sSleep)
+    {
+        PHY_Wakeup();
+    }
+    else if (sRxEnable)
+    {
+        PHY_SetRxState(false);
+    }
+}
+
+static void radioRestore()
+{
+    if (sSleep)
+    {
+        PHY_Sleep();
+    }
+    else if (sRxEnable)
+    {
+        PHY_SetRxState(true);
+    }
+}
+
+static void setTxPower(uint8_t aPower)
+{
+    if (aPower != sPower)
+    {
+        uint8_t i;
+
+        for (i = 0; i < sizeof(sTxPowerTable); i++)
+        {
+            if (aPower >= sTxPowerTable[i])
+            {
+                i++;
+                break;
+            }
+        }
+
+        otLogDebgPlat(sInstance, "Radio set tx power: %d, %d", aPower, i - 1);
```

Comment:
You are doing `i++` but then always referencing `i` as `i - 1`, why not just drop the increment and use `i` directly?

Reply:
It is for case when `aPower` is less than last element in `sTxPowerTable`. In this case `i == sizeof(sTxPowerTable)` after loop. I can do loop to `size - 1` and use `i` directly.